src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java

```diff
@@ -23,85 +33,76 @@
 import org.infinispan.commons.util.concurrent.CompletableFutures;
 
-import javax.annotation.PostConstruct;
-import javax.enterprise.context.ApplicationScoped;
-import javax.enterprise.event.Observes;
-import javax.inject.Inject;
-
-import java.util.Map;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.CompletionStage;
-import java.util.concurrent.ConcurrentHashMap;
-
 @ApplicationScoped
 public class HotRodService implements CacheService {
    private static final String OFFSET_CACHE_NAME = "___debezium-offset";
    private static final String SCHEMA_CACHE_NAME = "___debezium-schema";
 
    @Inject NotificationManager eventing;
 
-   @Inject Configuration config;
-
    @Inject DBSyncerMetrics metrics;
 
-   volatile RemoteCacheManager rcm;
-
-   private final Map<String, CacheBackend> backends = new ConcurrentHashMap<>();
+   private final Map<URI, HotRodCache> managers = new ConcurrentHashMap<>();
    private volatile boolean shutdownSignal;
 
-   @PostConstruct
-   public void init() {
-      rcm = new RemoteCacheManager(config.cache().uri());
-      createCaches();
-   }
-
    public void shutdown(@Observes ShutdownEvent ignore) {
-      if (rcm != null) {
-         shutdownSignal = true;
-         if (backends.values().stream().noneMatch(CacheBackend::isRunning)) destroy();
-      }
+      shutdownSignal = true;
+      destroy();
    }
 
    @Override
-   public CompletionStage<Boolean> reconnect(String name, Rule rule) {
-      if (!backends.containsKey(name)) return CompletableFutures.completedFalse();
-
-      return CompletableFuture.supplyAsync(() -> {
-         backendForRule(name, rule).start();
-         return null;
-      }).handle((ignore, t) -> {
-         if (t != null) {
-            eventing.backendFailedEvent(name, t);
-            return false;
-         }
-         createCaches();
-         return true;
+   public CompletionStage<Boolean> reconnect(CacheIdentifier identifier, Rule rule) {
+      var cache = getCacheOrThrow(identifier);
+      return cache.rcm.startAsync().thenApply(ignore -> {
+         var backend = getBackendOrThrow(identifier);
+         return backend.reconnect();
       });
    }
 
    @Override
-   public void stop(String name) {
-      if (!backends.containsKey(name)) throw new IllegalStateException(String.format("Backend for %s not created", name));
-      backends.computeIfPresent(name, (ignore, cache) -> {
-         cache.stop();
-         return cache;
-      });
+   public void stop(CacheIdentifier identifier) {
+      var cache = getCacheOrThrow(identifier);
+      var backend = cache.backends.remove(identifier.rule());
+
+      if (backend != null) backend.stop();
 
-      if (shutdownSignal) destroy();
+      destroy();
    }
 
    private void destroy() {
-      CompletionStage<Void> cs = rcm.stopAsync();
-      rcm = null;
+      if (!shutdownSignal || managers.values().stream().anyMatch(c -> !c.backends.isEmpty()))
```

Comment:
Looks like it doesn't shut down the manager if there are no more rules using the same uri?